examples/vision/simsiam.py

```diff
@@ -0,0 +1,431 @@
+"""
+Title: Self-supervised contrastive learning with SimSiam
+Author: [Sayak Paul](https://twitter.com/RisingSayak)
+Date created: 2021/03/19
+Last modified: 2021/03/20
+Description: Implementation of a self-supervised learning method for computer vision.
+"""
+"""
+Self-supervised learning (SSL) is an interesting branch of study in the field of
+representation learning. SSL systems try to formulate a supervised signal from a corpus
+of unlabeled data points.  An example is we train a deep neural network to predict the
+next word from a given set of words. In literature, these tasks are known as *pretext
+tasks* or *auxiliary tasks*. If we [train such a network](https://arxiv.org/abs/1801.06146) on a huge dataset (such as
+the [Wikipedia text corpus](https://www.corpusdata.org/wikipedia.asp)) it learns very effective
+representations that transfer well to downstream tasks. Language models like
+[BERT](https://arxiv.org/abs/1810.04805), [GPT-3](https://arxiv.org/abs/2005.14165),
+[ELMo](https://allennlp.org/elmo) all benefit from this.
```

Comment:
Suggestion:

- Describe the example in the first few sentences (and then later describe the steps, as you have already) (SimSiam on CIFAR-10).
- Shorten a few sentences (be mindful of the users attention span) without affecting the quality
- Some refactoring
- Link ELMo to https://arxiv.org/abs/1802.05365

For example:

```suggestion
Self-supervised learning (SSL)—a branch of study in representation learning—
tries to formulate a supervised signal from a corpus of unlabeled data points.
This example demonstrates how to implement one such SSL system—
simple Siamese networks (SimSiam) 
([Chen & He, 2020](https://arxiv.org/abs/2011.10566))—on the
[CIFAR-10](https://www.cs.toronto.edu/~kriz/cifar.html) dataset.
SSL can be applied to natural language modeling tasks, such as predicting the
next word from a given set of words, which is also known as *pretext tasks* or
*auxiliary tasks*. If we train a neural network, such as ULMFiT 
([Howard & Ruder, 2018](https://arxiv.org/abs/1801.06146)),
on the [Wikipedia text corpus](https://www.corpusdata.org/wikipedia.asp)),
the model can learn effective representations that can transfer well to
downstream tasks. Notable beneficiaries of this methodology are language models
like BERT ([Devlin et al., 2018](https://arxiv.org/abs/1810.04805)), GPT-3
([Brown et al., 2020](https://arxiv.org/abs/2005.14165)), and ELMo
([Peters et al., 2018](https://arxiv.org/abs/1802.05365)).
```